prov/rxm/src/rxm_cq.c

```diff
@@ -424,41 +424,76 @@ static ssize_t rxm_handle_seg_data(struct rxm_rx_buf *rx_buf)
 	return ret;
 }
 
-static ssize_t
-rxm_prepare_deferred_rndv_read(struct rxm_deferred_tx_entry **def_tx_entry,
-			       size_t index, struct iovec *iov,
-			       void *desc[RXM_IOV_LIMIT], size_t count,
-			       struct rxm_rx_buf *rx_buf)
+static ssize_t rxm_rndv_xfer(struct rxm_ep *rxm_ep, struct fid_ep *msg_ep,
+			     struct rxm_rndv_hdr *remote_hdr, struct iovec *local_iov,
+			     void **local_desc, size_t local_count, size_t total_len,
+			     void *context)
 {
-	uint8_t i;
+	size_t i, index = 0, offset = 0, count, copy_len;
+	struct iovec iov[RXM_IOV_LIMIT];
+	void *desc[RXM_IOV_LIMIT];
+	ssize_t ret = FI_SUCCESS;
 
-	*def_tx_entry = rxm_ep_alloc_deferred_tx_entry(rx_buf->ep, rx_buf->conn,
-						       RXM_DEFERRED_TX_RNDV_READ);
-	if (!*def_tx_entry)
-		return -FI_ENOMEM;
+	for (i = 0; i < remote_hdr->count && total_len > 0; i++) {
+		copy_len = MIN(remote_hdr->iov[i].len, total_len);
+
+		ret = ofi_copy_iov_desc(&iov[0], &desc[0], &count,
+					&local_iov[0],
+					&local_desc[0],
+					local_count,
+					&index, &offset, copy_len);
+		if (ret)
+			return ret;
+		total_len -= copy_len;
+		ret = rxm_ep->rndv_ops->xfer(msg_ep, iov, desc, count, 0,
+			       remote_hdr->iov[i].addr, remote_hdr->iov[i].key,
+			       context);
 
-	(*def_tx_entry)->rndv_read.rx_buf = rx_buf;
-	(*def_tx_entry)->rndv_read.rma_iov.addr =
-			rx_buf->remote_rndv_hdr->iov[index].addr;
-	(*def_tx_entry)->rndv_read.rma_iov.key =
-			rx_buf->remote_rndv_hdr->iov[index].key;
+		if (ret) {
+			if (ret == -FI_EAGAIN) {
+				struct rxm_deferred_tx_entry *def_tx_entry;
+
+				ret = rxm_ep->rndv_ops->defer_xfer(
+					&def_tx_entry, i, iov, desc, count,
+					context);
 
-	for (i = 0; i < count; i++) {
-		(*def_tx_entry)->rndv_read.rxm_iov.iov[i] = iov[i];
-		(*def_tx_entry)->rndv_read.rxm_iov.desc[i] = desc[i];
+				if (ret)
+					break;
+				rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
+				continue;
```

Comment:
my understanding of the original read code was to allow out of order chunks to be read, based on the sizing of the memory regions in play. Continuing allows this by continuing to chunk the message as it normally would and either immediately send or defer the chunk.

Reply:
If I understand the code, the chunking is posting 1 read/write for each iov entry.  But when we fail, we call defer_xfer, which maps to rxm_prepare_deferred_rndv_write/rxm_prepare_deferred_rndv_read.

Both of those functions will save the iov up to the ith entry.  The first loop, we'll only save iov[0].  The second time, we'll save iov[0] and iov[1], even though iov[0] was already saved.  The next loop, iov[0] ... iov[2], etc.

Eventually the deferred entries will be processed, which I think will result in reading/writing to the same original iov entry multiple times.  This increases the chances of future requests getting EAGAIN, which results in more backlog.

Reply:
Ok, I understand your point and agree this seems wrong - but functional.  I will work on clarifying this and reducing redundant copies.

Reply:
I'm not even sure about functional now.  The code likely completes the rndv protocol once the first deferred entry completes.  Which would result in the second deferred entry accessing memory that it shouldn't be.

Reply:
upon read (or write) completion, an rma index is incremented and compared against the expected rma count and does not complete until the count is reached, so that MR will not be released until all scheduled rma actions complete.  I don't think the second deferred entry would be accessing unintended memory.

Reply:
I see.  The code is assuming that the number of reads posted will equal the number of remote iov's.  It's creating a local iov that corresponds to a single remote iov that's being read.

The original code may be okay here, just very confusing.  iov is modified each iteration to reference the local recv iov.